libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
-    )
+    __cp_async_fallback<_Align>(__group, __dest_char, __src_char, __size);
+    return __completion_mechanism::__sync;
+}
 
-    return __is_async;
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
+__completion_mechanism __dispatch_memcpy_async_device(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    // Dispatch based on direction of the copy: global to shared, shared to
+    // global, etc.
+
+    // CUDA compilers < 12.3 may not propagate assumptions about the state space
+    // of pointers correctly. Therefore, we
+    // 1) put the code for each copy direction in a separate function, and
+    // 2) make sure none of the code paths can reach each other by "falling through".
+    //
+    // See nvbug 4074679 and also PR #478.
+    if (__isGlobal(__src_char) && __isShared(__dest_char)) {
+        return __dispatch_memcpy_async_global_to_shared<_Align>(__group, __dest_char, __src_char, __size, __allowed_completions, __bar_handle);
+    } else {
+        return __dispatch_memcpy_async_any_to_any<_Align>(__group, __dest_char, __src_char, __size, __allowed_completions, __bar_handle);
+    }
 }
 
+// __dispatch_memcpy_async is the internal entry point for dispatching to the correct memcpy_async implementation.
+template<typename _Group, typename _Tp, typename _Size>
+inline __host__ __device__
+__completion_mechanism __dispatch_memcpy_async(_Group const & __group, _Tp * __destination, _Tp const * __source, _Size __size, _CUDA_VSTD::uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    // When compiling with NVCC and GCC 4.8, certain user defined types that _are_ trivially copyable are
+    // incorrectly classified as not trivially copyable. Remove this assertion to allow for their usage with
+    // memcpy_async when compiling with GCC 4.8.
+    // FIXME: remove the #if once GCC 4.8 is no longer supported.
+#if !defined(_LIBCUDACXX_COMPILER_GCC) || _GNUC_VER > 408
+    static_assert(_CUDA_VSTD::is_trivially_copyable<_Tp>::value, "memcpy_async requires a trivially copyable type");
+#endif
+
+    // Alignment: Use the maximum of the alignment of _Tp and that of a possible cuda::aligned_size_t.
+    constexpr _CUDA_VSTD::size_t __size_align = __get_size_align<_Size>::align;
+    constexpr _CUDA_VSTD::size_t __align = (alignof(_Tp) < __size_align) ? __size_align : alignof(_Tp);
+
+    // Cast to char pointers. We don't need the type for alignment anymore and
+    // erasing the types reduces the number of instantiations of down-stream
+    // functions.
+    char * __dest_char = reinterpret_cast<char*>(__destination);
+    char const * __src_char = reinterpret_cast<char const *>(__source);
+
+    static_assert(_CUDA_VSTD::is_trivially_copyable<_Tp>::value, "memcpy_async requires a trivially copyable type");
```

Comment:
That assert is the same as above but unguarded

Reply:
Missed this. Thanks. Removed now.